File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int i = 0;
        int j = height.length-1;

        int maxLeft = 0;
        int maxRight = 0;
        int water = 0;

        while(i < j){
            if(height[i] <= height[j]){
                maxLeft = Math.max(height[i], maxLeft);
                water += maxLeft - height[i];
                i++;
            }else{
                maxRight = Math.max(height[j], maxRight);
                water += maxRight - height[j];
                j--;
            }
        }
        return water;
    }
}

File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution(object):
    def trap(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        i = 0
        j = len(height) - 1

        water = 0
        max_left, max_right = 0, 0

        while i < j:
            if height[i] <= height[j]:
                max_left = max(height[i], max_left)
                water += max_left - height[i]
                i += 1
            else:
                max_right = max(height[j], max_right)
                water += max_right - height[j]
                j -= 1
        return water
